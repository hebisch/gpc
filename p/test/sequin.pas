{ FLAG --field-widths=10 }

program initable(input,output);
type  counter=0..4;
      xobject=1..4;
      container=set of xobject;
      row=array[1..4]of 1..4;
var   sol:row; i,j,k:integer;
      table:array[1..24]of row;

    procedure select(bag:container; sample,drawn:counter);
    {Generation of permutations in lexicographic order,
    adapted from Grogono, page 134.}
    var  ball:xobject;
    begin
    if drawn<sample then
        for ball:=1 to 4 do
            if ball in bag then begin sol[drawn+1]:=ball;
                                select(bag-[ball],sample,drawn+1)
                                end;
    if drawn=sample then begin table[k]:=sol; k:=k+1 end
    end; {of select}

begin {of initable}
    k:=1;
    select([1..4],4,0);
    for i := 1 to k-1 do begin
        write('table[', i:1, '] = ');
        for j := 1 to 4 do
            write(table[i][j]);
        writeln;
        end;
end.
