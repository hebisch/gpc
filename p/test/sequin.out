table[1] =          1         2         3         4
table[2] =          1         2         4         3
table[3] =          1         3         2         4
table[4] =          1         3         4         2
table[5] =          1         4         2         3
table[6] =          1         4         3         2
table[7] =          2         1         3         4
table[8] =          2         1         4         3
table[9] =          2         3         1         4
table[10] =          2         3         4         1
table[11] =          2         4         1         3
table[12] =          2         4         3         1
table[13] =          3         1         2         4
table[14] =          3         1         4         2
table[15] =          3         2         1         4
table[16] =          3         2         4         1
table[17] =          3         4         1         2
table[18] =          3         4         2         1
table[19] =          4         1         2         3
table[20] =          4         1         3         2
table[21] =          4         2         1         3
table[22] =          4         2         3         1
table[23] =          4         3         1         2
table[24] =          4         3         2         1
